architecture-decision-records/Greenhouse-ADR-7-plugin-overrides.md

```diff
@@ -0,0 +1,92 @@
+# ADR-7 Plugin Option overrides
+
+## Decision Contributors
+
+- ...
+
+## Status
+
+- Proposed
+
+## Context and Problem Statement
+
+In Greenhouse Plugins are the primary way to extend the functionality of the Operations Platform. Since there are some Plugins that are required in most clusters, such as `CertManager`, there are PluginPresets. These PluginPresets are a way to define a default configuration for a Plugin, which is deployed to all clusters matching the PluginPreset's selector.
+
+The issue now is that there are some cases where the default configuration between two clusters is different, or requires a cluster-specific secret. This is currently not possible with PluginPresets, as they are applied with the same configuration to all clusters matching the selector.
+
+Another issue is setting default values that are valid for all plugins inside of an Organization, or for all plugins for a specific cluster. Currently, this requires setting these values in every Plugin's spec.
+
+Greenhouse should offer a way to override PluginOptionValues for a specific cluster, for all Plugins of a certain PluginDefinition, or for all plugins in an Organization.
+
+## Decision Drivers
+
+- Stability:
+  - Overrides should be consistent
+  - Overrides should be applied in a deterministic way (most specific last)
+  - There should be no conflicts between overrides or constant reconciliation loops
+  - Changes to the overrides should be applied to all relevant plugins
+
+- Transparency:
+  - End-users should be able to see/understand which overrides are applied to a Plugin
+
+- Compatibility:
+  - Overrides should be compatible with existing Plugins
+  - Overrides should be compatible with existing PluginPresets
+
+## Decision
+
+We will introduce a new CRD called `PluginOverride`. This CRD will allow users to override PluginOptionValues.
+It will be possible to:
```

Comment:
The `PluginOverride.Spec.Overrides` will specify the overrides. An override consists of a

- `path`: This matches the `Plugin.Spec.OptionValues[].Name` to be overrridden. _(Really? How do we make sure, no duplicate names exist if we do *NOT* specify a `PluginDefinition`?_
- `value`: The value to insert.

Reply:
> path: This matches the Plugin.Spec.OptionValues[].Name to be overrridden. (Really? How do we make sure, no duplicate names exist if we do NOT specify a PluginDefinition?

Fair point, the path in Overrides should be unique. The idea behind making the PluginDefinition optional was to make it easier/convenient to setting values across an org. I am open to making this an optional feature for now. It might get interesting once we decide on common labels enforced throughout Greenhouse. These could also bring a convention on how to declare them for Plugins, as to making the Plugin development process easier.